### DISCUSSION TITLE
Possible Mistake in Avalonia Custom Control DataContext Initialized with parent's instead of child's DataType

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#12790] [General]
I'm having a weird issue where Avalonia is setting a child's `DataContext` using the parent's `DataType` instead of the child's.

It would make sense for me that the default behavior when a DataContext isn't specified in xaml would be to use the DataType specified in the xaml file to instantiate the right ViewModel and feed the `DataContext`. But for some reasons, this isn't the case.

Is this intended? And in the case it is, why?
In the case that this wasn't intended, i'll create a proper issue with an simpler example than what i'll link next.

PS: Currently encountering this issue in a simple game i'm trying to port from Godot, hre is the [link](https://github.com/Mrcubix/Color-Tiles.Avalonia)

**Comment by timunie** [comment]
In your sample, can you show a link to an XAML file that has the mentioned issue? Or as an alternative, a more minimal sample.

**Reply by Mrcubix** [reply]
https://github.com/Mrcubix/Color-Tiles.Avalonia/blob/93c01b8accb07903112fe9fea1e3a2d404d66771/ColorTiles/Views/MainView.axaml#L16

You can see here that i don't provide a DataContext to both custom controls as i expect Avalonia to provide the correct one, but if you run it and try to click any of the buttons, it just crash trying to cast the parent's DataContext to whatever control's DataContext should be.

The custom control is fed the parent's DataContext instead of its DataType.
What's even more surprising is that Avalonia doesn't seem to crash at compile time or runtime, even if compiled bindings are enabled globally by default and the DataContext's DataType is different from the axaml.

Take a look at the variable on Watch on the left, being, the DataContext.
![Image](https://media.discordapp.net/attachments/912645078089080843/1148663486193336362/image.png)

**Reply by Mrcubix** [reply]
Any news on that? unable to find where Avalonia is feeding this one DataContext from, tried using the repo as a submodule instead of using the package to troubleshoot the issue.
Doesn't seem like it's fed in a conventional way.

update: can't seem to find how it's done in the generated code either

**Reply by timunie** [reply]
Do you mind having a different solution that doesn't rely un event handlers added and removed on DataContextChanged?

**Reply by timunie** [reply]
Tbh, I'd work with DataTemplates and Commands if possible. That way you can better maintain the game later on imo. I tried to compile your App, but it doesn't even compile for me as it says an image cannot be loaded.

**Comment by kekekeks** [comment]
Controls always inherit the data context from their parent. `x:DataType` is a hint for compiled binding system, it doesn't enforce anything and doesn't magically change the data context type.
The application author are responsible for providing the correct data context for child elements should they require one. You can do that via something like `<local:MyControl DataContext="{Binding PropertyInParentDataContext}" />`

**Reply by Mrcubix** [reply]
Guess i will add that to "things that doesn't make sense and never tell its user about it with a warning list", alongside StyledProperty among other things

**Reply by kekekeks** [reply]
The way DataContext inheritance works and StyledProperty (aka DependencyProperty) infrastructure match all other XAML frameworks. It's been like this since 2007 ¯\_(ツ)_/¯
